### PR TITLE
Update rc-local.md

### PR DESCRIPTION
**Pull request by epoch1970**
Say how old and feeble rc.local is
Refrain from using rc.local to run daemons
Refrain from needlessly using sudo in rc.local

### CODE REVIEW

**Comment by andrum99** [comment]
I think a better solution would be to just point the user to systemd - ideally a page in this repo. I don't think there is any mileage in adding more documentation about rc.local - as you note it doesn't always work as intended, and may disappear completely in future.

**Comment by epoch1970** [comment]
That's a bit drastic I think. You don't really need a systemd unit to display "My IP address is 1.2.3.4". 
But I wrote this because I see forum users using rc.local a lot, and with sudo on top of that.

**Comment by andrum99** [comment]
> That's a bit drastic I think. You don't really need a systemd unit to display "My IP address is 1.2.3.4".

I agree, but Ubuntu doesn't ship an rc.local file by default on Pi.

**Comment by lurch** [comment]
Where does Ubuntu come into things? :confused:  The documentation in this repo _only_ covers Raspberry Pi OS.

**Comment by andrum99** [comment]
> Where does Ubuntu come into things? 😕 The documentation in this repo _only_ covers Raspberry Pi OS.

You're right - that was irrelevant. My mistake.

**Comment by andrum99** [comment]
I think I have a solution. Previously `rc.local` was guaranteed to run as the last thing after all services have started. All `rc.local` currently does on Pi is print out the Pi's IP address. It would actually be useful if we could guaranteed this was shown immediately above the login shell. And folk want to run stuff easily on startup without wanting to worry up service dependencies.

The solution is simply to modify the `systemd` `rc_local.service` so that it always runs after all other services have been started up. That way we can still direct users to `rc.local`, and the IP address is shown where it was before systemd got its grubby paws on it. If we simply tell systemd to start rc_local.service after `multi-user.target` or `graphical.target` we get the expected behaviour. (Actually the IP address isn't printed to the console before X starts, but in console messages are disabled by default anway, and the first run wizard reports the IP address too).

**Comment by lurch** [comment]
I'm afraid I know very little about systemd, so pinging @XECDesign in case he has any comments / opinions.

**Comment by epoch1970** [comment]
As far as I know, “last” isn’t part of systemd’s vocabulary.
Besides I don’t think users have a problem with how rc.local works (or not). They just stuff it and expect that to work. Rc.local seems to be called fashionably late by systemd already.
Perhaps the rewrite will cause cron to get abused more than it is today... But at least some users might recognize systemd is the way to go for them.

**Comment by XECDesign** [comment]
I'm on the same page as @epoch1970. I'm not a fan of spending time maintaining rc.local. It is what it is with the limitations. In the ideal world, we'd have a page on writing systemd units (if we don't already) and this page would advice people to go there instead.

**Comment by andrum99** [comment]
> I'm on the same page as @epoch1970. I'm not a fan of spending time maintaining rc.local. It is what it is with the limitations. In the ideal world, we'd have a page on writing systemd units (if we don't already) and this page would advice people to go there instead.

Thanks for your input. My suggestion turned out to not really be feasible, and I think I can write a page for systemd quite easily, so will do that. I will leave it up to @epoch1970 to decide whether to continue with this PR, but expect my PR to suggest removal of most of the content on this page.

**Comment by epoch1970** [comment]
> I think I can write a page for systemd quite easily, so will do that.
https://github.com/raspberrypi/documentation/blob/master/linux/usage/systemd.md (the rc.local page does already link to it)
> I will leave it up to @epoch1970 to decide whether to continue with this PR, but expect my PR to suggest removal of most of the content on this page.
I can't decide on that. I did that PR because I think it might help users. Not massive numbers for sure. The current page is not bad. My bet is rc.local will still be around for a while, ppl will create the file if it's not there by default...

**Comment by lurch** [comment]
> My bet is rc.local will still be around for a while, ppl will create the file if it's not there by default...

Yeah, I suspect there's probably loads of (now outdated) 3rd-party RPi tutorials and HOWTOs floating around on the internet, which people will still be following, advising people to add commands to `/etc/rc.local` ... :confused:

**Comment by MichaIng** [comment]
I also suggest to not use and promote `rc.local` anymore nowadays but instead consider to remove it completely on new Raspberry Pi OS images. If really wanted, it can still be created, the `rc-local.service` automatically executes it when present and executable, else does nothing.

As mentioned it is unreliable and the exact execution time/order cannot be controlled. Due to its service type (better the `rc-local.service` service type) `idle` runs it when the systemd boot order is idle for a while, usually when all enabled services have started up. The local getty service is of the same type, so by chance either the getty prompt is first, leading to `rc.local` output overriding the user input prompt (one needs to hit return to get it again) or it is not visible at all because the getty prompt usually clears or at least moves up the terminal history.

A systemd unit is indeed overkill to print the IP around login. The following works much better anyway to show the IP nicely on the login prompt:
```
mkdir -p /etc/issue.d
echo 'My IP address is \4
' > /etc/issue.d/ip.issue
```
ip

**Comment by andrum99** [comment]
> I also suggest to not use and promote `rc.local` anymore nowadays but instead consider to remove it completely on new Raspberry Pi OS images. If really wanted, it can still be created, the static `rc-local.service` automatically executes it when present and executable, else does nothing.
> 
> As mentioned it is unreliable and the exact execution time/order cannot be controlled. Due to its service type (better the `rc-local.service` service type) `idle` runs it when the systemd boot order is idle for a while, usually when all enabled services have started up. The local getty service is of the same type, so by chance either the getty prompt is first, leading to `rc.local` output overriding the user input prompt (one needs to hit return to get it again) or it is not visible at all because the getty prompt usually clears or at least moves up the terminal history.
> 
> A systemd unit is indeed overkill to print the IP around login. The following works much better anyway to show the IP nicely on the login prompt:
> 
> ```
> mkdir -p /etc/issue.d
> echo 'My IP address is \4
> ' > /etc/issue.d/ip.issue
> ```
> 
> ip

If you can prove that works just as reliably as the line in rc.local, then please file a PR for the removal of rc.local, plus the addition of that change. There's no reason rc.local should still be there.

Am I right in saying that /etc/issue thing shows up on remote connections too?

**Comment by XECDesign** [comment]
> ```
> mkdir -p /etc/issue.d
> echo 'My IP address is \4
> ' > /etc/issue.d/ip.issue
> ```

I like it... nice and simple.

**Comment by andrum99** [comment]
> Am I right in saying that /etc/issue thing shows up on remote connections too?

It appears not.

Also, I've had a quick read, and the current code in /etc/rc.local will print the IP address of all interfaces. The good news is you can just insert the same code into an `issue` file.

**Comment by MichaIng** [comment]
Indeed `/etc/issue`(.d/*.issue) is first of all only shown on `agetty` login prompts. Everything else depends on the login handler, i.e. the SSH server. OpenSSH has the `Banner` option to show a custom entry: https://manpages.debian.org/buster/openssh-server/sshd_config.5.en.html#Banner
`/etc/issue.net` seems to be a convention for this which. But OpenSSH does not expand `\4` and other such codes, also I don't see a need to show the local server IP on remote SSH client, where you anyway need to connect with IP or domain name?
```
login as: root
Pre-authentication banner message from server:
Debian GNU/Linux 10 \4
End of banner message from server
root@192.168.1.22's password:
```
- Shown after entering username
- No escape code support
- IP/hostname used to connect shown on password input and at least PuTTY window title

`/etc/rc.local` output of course was never shown on SSH sessions as well, but as every systemd unit prints to local console at best, so in this regards it is no regression 😉.

> Also, I've had a quick read, and the current code in /etc/rc.local will print the IP address of all interfaces

That is true: https://github.com/RPi-Distro/pi-gen/blob/master/stage2/01-sys-tweaks/files/rc.local
It picks up the first configured/active interface and prints its first IP: https://manpages.debian.org/buster/util-linux/agetty.8.en.html#ISSUE_FILES

One could add `\6` or `\4{wlan0}` but if e.g. IPv6 is disabled it will still show `::ffff:127.0.1.1` (derived from `/etc/hosts` localhost entry, I guess) and selecting a specific interface which might not exist is as well not clean, `\4 \4{wlan0` might double wlan0 IP if no Ethernet is connected...

> The good news is you can just insert the same code that's in rc.local into an issue file.

Nope, from what I know this is not possible, no shell code or execution possible/interpreted in issue files. One way to get/keep the old content is to add a systemd unit which rewrites `/etc/issue.d/ip.issue`:
```
cat << _EOF_ > /etc/systemd/system/update-login-info.service
[Unit]
Wants=network-online.target
After=network-online.target

[Service]
Type=oneshot
ExecStart=/bin/sh -c '_IP=\$(hostname -I); [ "\$_IP" ] && printf "My IP is \$_IP\n\n" > /etc/issue.d/ip.issue || true'

[Install]
WantedBy=multi-user.target
_EOF_
systemctl daemon-reload
systemctl enable --now update-login-info
systemctl restart getty@tty1
```
Although for multiple IPs it should be "My IPs are ..." then.

Basically the use case for this feature is essential. I just remember a statement where a guy had several SBCs attached to a HDMI switch so he was able to quickly switch between multiple SBC TTYs and the having the local IP info was a way to assure/identify which SBCs TTY he is currently attached to. In this case the first local v4 IP was pretty fine, but not sure if for other reasons users might need to see all IPs of all interfaces.

**Comment by andrum99** [comment]
The aim with any replacement to the current printing of the IP address(es) in rc.local would be to not print them on remote connections, to match the current implementation, which is why I was asking. It looks like rc.local will need to be left in place for the IP address thingy. I will knock up a page documenting how to write a systemd unit file, and what some of the useful options might be for noobs.

**Comment by epoch1970** [comment]
> `/etc/issue.net` seems to be a convention for this. But OpenSSH does not expand `\4` and other such codes, also I don't see a need to show the local server IP on remote SSH client, where you anyway need to connect with IP or hostname?

I think your initial suggestion works perfectly well iff an empty (or \4-less) issue.net is created as well.
I'll experiment later on to confirm that.

IMHO the subject is perpendicular to documenting how to use rc.local. Again I am convinced that if removed from the base image, and users will bring it back. And misuse it.

**Comment by MichaIng** [comment]
> The aim with any replacement to the current printing of the IP address(es) in rc.local would be to not print them on remote connections, to match the current implementation, which is why I was asking.

In this regards, `/etc/issue.d` works perfectly fine.

> It looks like rc.local will need to be left in place for the IP address thingy.

It the output + behaviour must stay as it is currently, then yes, as the solution with `update-login-info.service` above is a bid too complication (IMO). But really, `rc.local` based output does not work fine either:
- `rc.local` is executed before `getty@tty1`, and due to default `TTYVTDisallocate=yes` the IP info is not visible: 
ip1
- On RPi to prevent this, `TTYVTDisallocate=no` override for `getty@tty1.service` is present: https://github.com/RPi-Distro/pi-gen/blob/master/stage1/01-sys-tweaks/files/noclear.conf
Now, if `rc.local` is not executed too early (difficult to control, depends on overall systemd units and order), it is visible, but only as part of all the other systemd output: 
ip2
When switching to a different TTY for default console or if one actually wants to have a clean login prompt without boot messages left, we're again on first picture.
- To make it cleaner, on RPi `quiet` cmdline entry is added by default: https://github.com/RPi-Distro/pi-gen/blob/master/stage2/01-sys-tweaks/00-patches/07-resize-init.diff
Now it is better, but still not beautiful (timestamp, PID): 
ip3
And again, if one changes settings e.g. because kernel output on boot is wanted for debug/informational reasons, we're again at the second picture, if there are more units started between those two, the info can be moved out of screen.

So it is at best fragile and based on two patches away from Debian default (compare code lines to `update-login-info.service` alternative 🤔) which both is something users might want to alter again. An `/etc/issue.d` addition is much cleaner and output is always at the same place regardless of above mentioned changes. In my personal opinion the benefits outweigh the regression of only one first IP being shown.

**Comment by MichaIng** [comment]
> I think your initial suggestion works perfectly well iff an empty (or \4-less) issue.net is created as well.

Note that, even I find `/etc/issue.net` present on fresh Debian installs (netboot installer), it is not used by default by OpenSSH. One needs to explicitly change default `sshd_config`, enable and set `Banner=/etc/issue.net`. And Dropbear cannot use it at all. So I would not bother or try to add or change anything about that, leave SSH as it always was: Clean prompt, no additional info, also for privacy reasons.

**Comment by andrum99** [comment]
> > The aim with any replacement to the current printing of the IP address(es) in rc.local would be to not print them on remote connections, to match the current implementation, which is why I was asking.
> 
> In this regards, `/etc/issue.d` works perfectly fine.

Except it doesn't. The issue file only reports the first IPv4 interface.

**Comment by MichaIng** [comment]
> Except it doesn't. The issue file only reports the first IPv4 interface.

"In this regards" meant that it will not print anything on remote connections. Yes it does not behave exactly the same, but has pros and cons as discussed in the same post. Decide whether the pros outweigh the cons or not.

**Comment by andrum99** [comment]
Given there isn't a better way to print all the IP addresses at boot, rc.local will need to stay. I will write the page on systemd.

**Comment by andrum99** [comment]
I recommend this PR be closed. See #1617.
